experimental/packages/opentelemetry-sdk-metrics-base/src/exemplar/ExemplarReservoir.ts

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { ValueType, Attributes } from '@opentelemetry/api-metrics'
+import { Context, HrTime, isSpanContextValid, trace } from '@opentelemetry/api'
+import { Exemplar } from './Exemplar'
+
+
+/**
+ * An interface for an exemplar reservoir of samples.
+ */
+export interface ExemplarReservoir {
+
+  /** Offers a measurement to be sampled. */
+  offer(
+    value: ValueType,
+    timestamp: HrTime,
+    attributes: Attributes,
+    ctx: Context    
+  ): void;
+  /**
+   * Returns accumulated Exemplars and also resets the reservoir
+   * for the next sampling period
+   * 
+   * @param pointAttributes The attributes associated with metric point. 
+   */
+  collectAndReset(pointAttributes: Attributes): Exemplar[];
+}
+
+
+class ExamplarBucket {
+  private value: ValueType = 0;
+  private attributes: Attributes = null;
+  private timestamp: HrTime = [0, 0];
+  private spanId?: string;
+  private traceId?: string;
+
+  constructor() {}
+
+  offer(value: ValueType, timestamp: HrTime, attributes: Attributes, ctx: Context) {
+    this.value = value;
+    this.timestamp = timestamp;
+    this.attributes = attributes;
+    const spanContext = trace.getSpanContext(ctx);
+    if (spanContext && isSpanContextValid(spanContext)) {
+      this.spanId = spanContext.spanId;
+      this.traceId = spanContext.traceId;
+    }
+  }
+
+  collectAndReset(pointAttributes: Attributes): Exemplar | null {
```

Comment:
I can't find logic about reseting reservoir in the specification, i am missing something ?

Reply:
Spec is really vague in terms of defining this. The relevant part is that we need to have a fixed size reservoir and when the exemplars are collected we reset them and use later when measurements are offered. Does it make sense?

Reply:
It make sense but i would say we don't need to expose the `reset` logic, can't we just have a `collect` method ?

Reply:
I agree and initially used just the `collect` but I update it based on this conversation https://github.com/open-telemetry/opentelemetry-js/pull/2641#discussion_r755648188.

Reply:
didn't see it mybad, i understand the point of legendecas but i'm not sure of what we want to do here (maybe ask for a spec clarification ?) , cc @dyladan

Reply:
+1 it is vague, and I think resetting every time is generally what you want.

However, since users may implement ExemplarReservoirs themselves, they need not reset them every collection if they don't want to. So my vote would be `collect()` and a good doc string

Reply:
FYI - This is a good callout.  There were a few clarifying things in the spec that got dropped, but we can add:


1. The collect+reset here is to account for DELTA vs. CUMULATIVE aggregation windows.   Exemplars collection should mirror aggregation.  In Java implementation, all synchronous instruments *always* report DELTA accumulations, hence collect+reset by default.
2. Exemplar preservation should be best-effort.   It might be worth calling out that's it's explicitly ok to drop exemplar memory storage under pressure.

IF you open a ticket for spec clarification I'll tackle it.

Reply:
We can't make this rely on the temporality until the aggregation PR is finished I think right?

Reply:
Which aggregation PR?   AFAIK everything you need is already in the Specification for this requirement.  Clarification PR pending.

Reply:
I mean the JS implementation of aggregations. we can't have an aggregation temporality without an aggregation. maybe i'm wrong and its already implemented